beacon-chain/p2p/gossipsub_peer_controller.go

```diff
@@ -0,0 +1,153 @@
+package p2p
+
+import (
+	"context"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/OffchainLabs/prysm/v7/beacon-chain/p2p/gossipsubcrawler"
+	"github.com/OffchainLabs/prysm/v7/cmd/beacon-chain/flags"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/pkg/errors"
+)
+
+const (
+	peerPerTopic = 20
+)
+
+type GossipsubPeerDialer struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	listPeersFunc func(topic string) []peer.ID
+	dialPeersFunc func(ctx context.Context, maxConcurrentDials int, nodes []*enode.Node) uint
+
+	crawler        gossipsubcrawler.Crawler
+	topicsProvider gossipsubcrawler.SubnetTopicsProvider
+
+	wg   sync.WaitGroup
+	once sync.Once
+}
+
+func NewGossipsubPeerDialer(
+	crawler gossipsubcrawler.Crawler,
+	listPeers func(topic string) []peer.ID,
+	dialPeers func(ctx context.Context, maxConcurrentDials int, nodes []*enode.Node) uint,
+) *GossipsubPeerDialer {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &GossipsubPeerDialer{
+		listPeersFunc: listPeers,
+		dialPeersFunc: dialPeers,
+		crawler:       crawler,
+		ctx:           ctx,
+		cancel:        cancel,
+	}
+}
+
+func (g *GossipsubPeerDialer) Stop() {
+	g.cancel()
+	g.wg.Wait()
+}
+
+func (g *GossipsubPeerDialer) Start(provider gossipsubcrawler.SubnetTopicsProvider) error {
+	if provider == nil {
+		return errors.New("topics provider is nil")
+	}
+
+	g.once.Do(func() {
+		g.topicsProvider = provider
+		g.wg.Go(func() {
+			g.dialLoop()
+		})
+	})
+
+	return nil
+}
+
+func (g *GossipsubPeerDialer) dialLoop() {
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			topics := g.topicsProvider()
+			var peersToDial []*enode.Node
+
+			for _, topic := range topics {
+				newPeers := g.peersForTopic(topic, peerPerTopic)
+				peersToDial = append(peersToDial, newPeers...)
+			}
+
+			if len(peersToDial) == 0 {
+				continue
+			}
+
+			// Deduplicate peers to avoid dialing the same peer multiple times.
+			uniquePeers := make([]*enode.Node, 0, len(peersToDial))
```

Comment:
Why creating a slice of peers, then deduplicating them with a set, then converting them back into a slice?
`g.dialPeers` could directly take a set of peers instead.